bin/plugin/commands/performance.js

```diff
@@ -200,23 +200,39 @@ function curateResults( testSuite, results ) {
  * @return {Promise<WPPerformanceResults>} Performance results for the branch.
  */
 async function runTestSuite( testSuite, performanceTestDirectory, runKey ) {
-	await runShellScript(
-		`npm run test:performance -- packages/e2e-tests/specs/performance/${ testSuite }.test.js`,
-		performanceTestDirectory
-	);
-	const resultsFile = path.join(
-		performanceTestDirectory,
-		`packages/e2e-tests/specs/performance/${ testSuite }.test.results.json`
-	);
-	fs.mkdirSync( './__test-results', { recursive: true } );
-	fs.copyFileSync( resultsFile, `./__test-results/${ runKey }.results.json` );
-	const rawResults = await readJSONFile(
-		path.join(
+	try {
+		await runShellScript(
+			`npm run test:performance -- packages/e2e-tests/specs/performance/${ testSuite }.test.js`,
+			performanceTestDirectory
+		);
+		const resultsFile = path.join(
 			performanceTestDirectory,
 			`packages/e2e-tests/specs/performance/${ testSuite }.test.results.json`
-		)
-	);
-	return curateResults( testSuite, rawResults );
+		);
+		fs.mkdirSync( './__test-results', { recursive: true } );
+		fs.copyFileSync(
+			resultsFile,
+			`./__test-results/${ runKey }.results.json`
+		);
+		const rawResults = await readJSONFile(
+			path.join(
+				performanceTestDirectory,
+				`packages/e2e-tests/specs/performance/${ testSuite }.test.results.json`
+			)
+		);
+		return curateResults( testSuite, rawResults );
+	} catch ( error ) {
+		fs.mkdirSync( './__test-results/artifacts', { recursive: true } );
+		const artifactsFolder = path.join(
+			performanceTestDirectory,
+			'artifacts'
+		);
+		await runShellScript(
+			'cp -Rv ' + artifactsFolder + ' ' + './__test-results/artifacts'
+		);
+
+		throw error;
+	}
```

Comment:
why should we only copy these if we fail? would we not want to upload them unconditionally so we can examine them even when the tests pass?

Reply:
AFAIK artifacts are not created for successful tests.

Reply:
aha, hm. I guess none of this would fail if we run it unconditionally? you're saying that the test runner doesn't create artifacts when the tests all pass?

just asking here because I've found it useful to analyze successful test results (e.g. the performance traces), but if they're not present then I guess there's no point.

the `try/catch` here is what holds me up. note that we're catching a lot of cases here, so for example, if the tests pass but `curateResults` fails, we'll still be copying the absence of files.

this is fine as it is, but I think it'd also be fine to leave the `try/catch` out of this and blanket copy all the artifacts, even if they don't exist. do you think there's a reason not to do that?

Reply:
> you're saying that the test runner doesn't create artifacts when the tests all pass?

The artifacts I'm trying to handle here are test failure ones, which include a page screenshot and HTML snapshot (at the moment of failure). This is different from the performance results json that we create only if the tests pass.

>the try/catch here is what holds me up. note that we're catching a lot of cases here, so for example, if the tests pass but curateResults fails, we'll still be copying the absence of files.
>
> this is fine as it is, but I think it'd also be fine to leave the try/catch out of this and blanket copy all the artifacts, even if they don't exist. do you think there's a reason not to do that?

I think we need this try/catch block here because otherwise, it would never get to the part where we copy the created artifacts. I agree, though, that it would be better to only wrap the bit where we run the tests (`npm run test:performance ...`). I'll make an update.

Reply:
https://github.com/WordPress/gutenberg/pull/48243/commits/fcf67c6ccb053c56fc289b750ba046c774474379